README.md
# transaction-bank

[How to use](https://github.com/romuloslv/transaction-bank/blob/master/doc/solution.md)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-07-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-07-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/transaction-bank/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/transaction-bank/compare/0.1.0...0.1.1


resources/request/report-test.sh
#!/bin/sh

TEST_PAGE="$(which google-chrome) --no-sandbox $PWD/target/coverage/index.html"

lein cloverage --runner :midje 2> /dev/null
$TEST_PAGE 2> /dev/null


test/transaction_bank/card_test.clj
(ns transaction-bank.card-test
  (:require [midje.sweet :refer :all]
            [cheshire.core :refer :all]
            [transaction-bank.card :refer :all]))

(fact "Test current-acumulator"
  (current-acumulator 6100 0 1) => 6100)

(fact "Test rules"
  (rules "true" 6100 ["Padaria" "Farmacia" "Mercado"]
         "Posto" 300 "2019-07-29 20:49:55") => [])

(fact "Find how many merchants exceed the rules limit"
  (loop [x 0]
    (when (< x 10)
      (moment "Uber" (.format
                     (java.text.SimpleDateFormat.
                          "yyyy-MM-dd HH:mm:ss")
                          (new java.util.Date)))
      (recur (+ x 1))))
  (comparation)  => ["Uber"]
  (reset! all-transactions []))

(fact "Recalculated limit basead in error"
  (current-acumulator 5800 1456 5) => 5800
  (current-acumulator 5800 1370 0) => 4430)

(fact "Recalculated limit"
  (def amount 430)
  (rescue acumulator amount) => 4000)

(fact "Test rules"
  (rules "false" 0 ["Padaria" "Farmacia" "Mercado"]
         "Farmacia" 20 "2019-07-29 20:49:55") => [])

(fact "Test rules"
  (rules "false" 10 ["Padaria" "Farmacia" "Mercado"]
         "Farmacia" 20 "2019-07-29 20:49:55") => [])

(fact "Test rules"
  (rules "false" 100 ["Padaria" "Farmacia" "Mercado"]
         "Farmacia" 20 "2019-07-29 20:49:55") => [])

(fact "Test limit merchant"
  (loop [x 0]
    (when (< x 10)
      (rules "true" 100 ["Padaria" "Farmacia" "Mercado"]
             "Posto" 20 "2019-07-29 20:49:55")
      (recur (+ x 1))))
  (rules "true" 100 ["Padaria" "Farmacia" "Mercado"]
         "Posto" 20 "2019-07-29 20:49:55")  => [])

(fact "Record test of a transaction"
  (reset! all-transactions [])
  (def out {:merchant "Padaria"
            :timestamp "2019-07-29 18:13:52"
            :millisecond (now)})
  (moment "Padaria" "2019-07-29 18:13:52") => [out]
  (reset! all-transactions []))

(facts "Tests defined by the execution rules"
  (fact "When the code is ten"
    (status-code 10)	=>	[{:error-body "it's all right."
                            :error-name "OK"}]
    (reset! erro ()))
  (fact "When the code is twenty"
    (status-code 20)	=>	[{:error-name "Unauthorized"
                            :error-body "Card Block in your APP."}]
    (reset! erro ()))
  (fact "When the code is thirty"
    (status-code 30)	=>	[{:error-name "Forbidden"
                            :error-body "Your request is authenticated but has insufficient limit."}]
    (reset! erro ()))
  (fact "When the code is forty"
    (status-code 40)	=>	[{:error-name "Method Not Allowed"
                            :error-body "You are using the card in a Blacklist establisment."}]
    (reset! erro ()))
  (fact "When the code is fifty"
    (status-code 50)	=>	[{:error-name "Buy Denied"
                            :error-body "Your buy is exceeding its limit."}]
    (reset! erro ()))
  (fact "When the code is sixty"
    (status-code 60)	=>	[{:error-name "First Buy"
                            :error-body "Your first purchase exceeds the 90% rate limit."}]
    (reset! erro ()))
  (fact "When the code is seventy"
    (status-code 70)	=>	[{:error-name "Merchant Limited"
                            :error-body "Merchant whith more of ten transactions."}]
    (reset! erro ()))
  (fact "When the code is eighty"
    (status-code 80)	=>	[{:error-name "Gateway Timeout"
                            :error-body "Many transactions in less than two minutes"}]
    (reset! erro ())))


test/transaction_bank/core_test.clj
(ns transaction-bank.core-test
  (:require [midje.sweet :refer :all]
    	      [ring.mock.request :as mock]
            [cheshire.core :refer :all]
            [transaction-bank.core :refer :all]))

(facts "Returns instructions on root route"
  (let [response	(application	(mock/request	:get	"/"))]
    (fact "Answer status is 200"
      (:status	response)	=>	200)
    (fact "Body text is an instruction"
      (:body	response)	=>	"*** Access the README to execute tests ***")))

(facts "Invalid route does not exist"
  (let [response	(application	(mock/request	:get	"/not-found"))]
    (fact "The error code is 404"
      (:status	response)	=>	404)
    (fact "The body text is 'Not Found!'"
      (:body response) => "Not Found!")))

(facts "Send data to service"
  (let [response (application (-> (mock/request :post "/transactions")
                                  (mock/json-body {:cardIsActive "true"
                                                   :limit 6100
                                                   :denylist ["Padaria" "Farmacia" "Mercado"]
                                                   :merchant "Mercado"
                                                   :amount 300
                                                   :timestamp "2019-07-29 20:49:55"})))]
    (fact "The code is 200"
      (:status	response)	=>	200)
    (fact ""
      (:body response) => "[]")))


project.clj
(defproject transaction-bank "0.1.0-SNAPSHOT"
  :description "A simple application for managing transactions"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.0"]
                 [org.clojure/clojurescript "1.10.191"]
                 [compojure "1.6.1"]
                 [ring/ring-json "0.4.0"]
                 [ring/ring-mock "0.3.2"]
                 [cheshire "5.8.1"]
                 [http-kit "2.3.0"]
                 [clj-time "0.14.0"]
                 [midje	"1.9.6"]]
  :plugins	[[lein-midje "3.2.1"]
             [lein-cloverage "1.0.13"]]
  :main ^:skip-aot transaction-bank.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


src/transaction_bank/core.clj
(ns transaction-bank.core
  (:require [org.httpkit.server :as server]
            [compojure.core :refer :all]
            [compojure.route :as route]
            [ring.middleware.json :as middleware]
            [transaction-bank.api :as api])
  (:gen-class))

(def error-route
  {:body (str "Not Found!")})

(defn msg-index
  []
  (str "*** Access the README to execute tests ***"))

(defroutes server-routes
  (GET "/" [] (msg-index))
  (POST "/transactions" {body :body} (api/request-transaction body))
  (route/not-found (:body error-route)))

(defn attach-middleware
  "Add Ring middleware to parse json request bodies and format them later"
  [app-routes]
  (-> app-routes
    (middleware/wrap-json-body)
    (middleware/wrap-json-response)))

(def application
  (attach-middleware server-routes))

(defn -main
  [& args]
  (println "*** Web server running on port 3003 ***")
  (server/run-server application {:port 3003}))


src/transaction_bank/card.clj
(ns transaction-bank.card
  (:require [clojure.java.io :as io]))

(def acumulator 0)

(def erro
  (atom ()))

(def all-transactions
  (atom []))

(defn now
  []
  (.getTime (java.util.Date.)))

(defn rescue
  "Change final value based on value received from current-accumulator"
  [args amount]
  (alter-var-root #'acumulator #(last [%1 (- args amount)])))

(defn moment
  "Saves the time at which a transaction was executed."
  [merchant timestamp]
  (swap! all-transactions conj {:merchant  merchant
                                :timestamp timestamp
                                :millisecond (now)}))

(defn comparation
  "Returns the amount of merchants with more than ten transactions"
  []
  (->> (mapv :merchant @all-transactions)
       sort
       (partition-by identity)
       (filter #(> (count %) 9))
       (mapv first)))

(defn current-acumulator
  "Calculates initial values ​​based on transaction-bank/resources/transactions.json file entry"
  [limit amount count-erro]
  (if (zero? count-erro)
    (if (> acumulator 0)
      (rescue acumulator amount)
      (rescue limit amount))
    (if (zero? acumulator)
      limit
      acumulator)))

(defn status-code
  [error-code]
  (cond
  (= error-code 10)
    (swap! erro conj {:error-name "OK"
                      :error-body "it's all right."})
  (= error-code 20)
    (swap! erro conj {:error-name "Unauthorized"
                      :error-body "Card Block in your APP."})
  (= error-code 30)
    (swap! erro conj {:error-name "Forbidden"
                      :error-body "Your request is authenticated but has insufficient limit."})
  (= error-code 40)
    (swap! erro conj {:error-name "Method Not Allowed"
                      :error-body "You are using the card in a Blacklist establisment."})
  (= error-code 50)
    (swap! erro conj {:error-name "Buy Denied"
                      :error-body "Your buy is exceeding its limit."})
  (= error-code 60)
    (swap! erro conj {:error-name "First Buy"
                      :error-body "Your first purchase exceeds the 90% rate limit."})
  (= error-code 70)
    (swap! erro conj {:error-name "Merchant Limited"
                      :error-body "Merchant whith more of ten transactions."})
  (= error-code 80)
    (swap! erro conj {:error-name "Gateway Timeout"
                      :error-body "Many transactions in less than two minutes"})
    :else (throw (Throwable. "An unknown error has occurred."))))

(defn output
  [limit merchant count-erro amount timestamp]
  (moment merchant timestamp)
  (if (zero? count-erro)
    (do
      (status-code 10)
      (def card-status true))
    (def card-status false))
  (with-open [wrtr (io/writer (str (get (System/getenv) "PWD")"/resources/output.json") :append true)]
    (.write wrtr (str "{""\n"
                      "   \"approved\": " "\"" card-status "\"," "\n"
                      "   \"newLimit\": " "\"" (current-acumulator limit amount count-erro) "\"," "\n"
                      "   \"deniedReasons\": " (mapv :error-body @erro) "\n"
                      "{""\n\n")))
  (reset! erro ()))

(defn rules
  "rules defined in transaction-bank/doc/intro.md"
  [card-status limit denylist merchant amount timestamp]
  (if (false? (read-string card-status))
    (status-code 20))

  (if (false? (> (bigdec limit) 0))
    (status-code 30))

  (if (.contains denylist merchant)
    (status-code 40))

  (if (or (>= (bigdec amount) (bigdec limit))
      (and (> acumulator 0) (> amount acumulator)))
    (status-code 50))

  (if (and (empty? @all-transactions)
           (false? (<= (bigdec amount) (* 0.9 (bigdec limit)))))
    (status-code 60))

  (if (.contains (comparation) merchant)
    (status-code 70))

  (if (> (count @all-transactions) 2)
    (if (<= (- (now) (first (mapv :millisecond @all-transactions))) 120000)
      (status-code 80)
      (reset! all-transactions [])))

  (output limit merchant (count @erro) amount timestamp))
